File: demo/django-proj/pages/views/study.py
# coding=utf-8

import random, json
from datetime import datetime
from django.utils.translation import ugettext_lazy as _

from config import *
from ..util import card_count_of

def study(request, author, deck_slug):
	# Initialize values that may or may not be used later
	heading    = ''
	time_taken = 0
	
	template            = 'study.html'
	title               = _('Now Studying') % {}
	content_class       = 'study-page'
	toolbar_class       = 'is-muted'
	minimalist_menubar  = True
	back_to_deck        = False
	cancel_back_to_deck = True
	
	
	
	# Get the deck object and total number of cards in this deck
	deck = Deck.objects.get(slug=deck_slug, author__name=author)
	card_count = request.POST.get('card-count', card_count_of(deck))
	
	
	
	# Determine whether a card was flipped
	card_was_flipped = request.POST.get('show-answer', False)
	
	# Determine whether we need to show the next card (this only occurs when a card wasn't flipped)
	show_next_card = request.POST.get('repeat') or request.POST.get('short-review-delay') or request.POST.get('medium-review-delay') or request.POST.get('long-review-delay') or request.POST.get('next')
	
	
	
	# Get and initialize our study options and data
	
	# Get the number of cards to offset by and the randomize seed (these are only set if JS is available)
	card_offset    = int(request.POST.get('offset', 0))
	randomize_seed = int(request.POST.get('randomize-seed', -1))
	
	# Get the start time and time options
	start_time = request.POST.get('start-time', False)
	use_time   = request.POST.get('use-time', '') # Boolean friendly string
	time_limit = float(request.POST.get('time-limit', 0))
	
	# Get the rest of the study options
	randomized   = request.POST.get('randomize', '') # Boolean friendly string
	random_queue = request.POST.get('random-queue')
	use_reverse  = request.POST.get('use-reverse', '') # Boolean friendly string
	repeat_queue = json.loads(request.POST.get('repeat-queue', '[]'))
	
	# If the time the user started studying hasn't been logged yet, set it now
	if not start_time:
		start_time = datetime.strftime(datetime.utcnow(), '%Y-%m-%d %H:%M:%S')
	
	
	
	# Track our progress in this session
	
	current_item = int(request.POST.get('current-item', 0))
	last_item    = int(request.POST.get('last-item', request.POST.get('cards-per-session', 20))) # Total number of cards to study in this session
	order = int(request.POST.get('order', card_offset + 1)) # The card we're currently displaying
	
	# Limit the number of cards to study by the total number in the deck
	if last_item > card_count:
		last_item = card_count
	
	
	
	# If we're randomizing and we don't have a random queue initialzed, create one and randomize the current card
	if randomized and not random_queue:
		random_queue = range(1, card_count+1)
		
		# If a seed was provided, seed the random number generator
		if randomize_seed != -1:
			random.seed(randomize_seed)
		
		random.shuffle(random_queue) # Shuffle the card order
		random_queue = random_queue[card_offset:] # Offset the values by our card offset value
		
		order = random_queue.pop()
	
	
	
	# Add this card to the repeat queue if the user chose to repeat it
	if request.POST.get('repeat'):
		repeat_queue.append(order)
	
	
	
	# Check for conditions that would terminate the study session
	
	# The session ended if the current card is also the last card and we have no cards to repeat
	session_ended = current_item == last_item and not repeat_queue
	# Whether or not time is up
	# @TODO Don't end the session on a time out if the user hasn't seen the answer to this card yet
	# @TODO This is an absolute mess, converting time strings into datetime objects and back and forth
	time_up = use_time and time_limit < (datetime.utcnow() - datetime.strptime(start_time, '%Y-%m-%d %H:%M:%S')).seconds
	
	
	
	# If the only thing left in the session is repeat cards, use the cards from the repeat queue
	if current_item == last_item and repeat_queue:
		# If we need to show a new card...
		if show_next_card:
			# ...get the value of that card's order, but don't remove it from the queue
			# This prevents the session from ending prematurely after the user chooses "show answer" on the last repeated card
			order = int(repeat_queue[0])
		# Otherwise, we're safe to remove the card from the queue
		else:
			repeat_queue.pop(0)
			
	
	
	# If the session ended
	if session_ended or time_up:
		# Randomize the page's heading
		headings = [
			_(u'That’s It!'),
			_(u'You’re Done!'),
			_('Congratulations!'),
			_('Nice Work!')
		]
		random.shuffle(headings)
		
		# Change the template to render, and other template-specific values
		heading             = headings.pop()
		template            = 'completion.html'
		title               = _('Studying Complete')%{}
		content_class       = 'completion-page'
		toolbar_class       = ''
		minimalist_menubar  = False
		back_to_deck        = True
		cancel_back_to_deck = False
		
		# Increment the deck's study count
		deck.times_studied += 1
		deck.save()
		
		# If time is up, show the time limit as the total time
		if time_up:
			total_time = time_limit
		# Otherwise show the amount of time the user spent studying
		else:
			start_time = datetime.strptime(start_time, '%Y-%m-%d %H:%M:%S')
			total_time = (datetime.utcnow() - start_time).seconds
		
		hours, remainder = divmod(total_time, 3600)
		minutes, seconds = divmod(remainder, 60)
		
		# Only show the hours place if the user took that long
		if hours:
			time_taken = '%d:%02d:%02d' % (hours, minutes, seconds)
		else:
			time_taken = '%d:%02d' % (minutes, seconds)	
	# Otherwise, check if we need to show the next card or mark the current one as completed if we're not on repeat cards
	elif current_item != last_item:
		# If a card was flipped, count the card as completed
		if card_was_flipped:
			current_item += 1
		# Otherwise, show the next card
		elif show_next_card:
			if randomized:
				random_queue = json.loads(random_queue)
				order = random_queue.pop()
			else:
				order += 1
	
	
	
	# If this is the first card...
	if current_item == 0:
		# Convert the time limit from minutes to seconds
		time_limit *= 60
	
	
	
	# Get the current card to display
	card = Card.objects.get(deck=deck, order=order)
	
	
	
	return render_to_response(template, {
		'title':   title + SEPARATOR + escape(deck.name) + SEPARATOR + SITE_TITLE,
		'heading': heading,
		
		'minimalist_menubar':  minimalist_menubar,
		'cancel_back_to_deck': cancel_back_to_deck,
		'back_to_deck':        back_to_deck,
		
		'toolbar_class': toolbar_class,
		'content_class': content_class,
		
		'deck': deck,
		'card': card,
		
		'card_count': card_count,
		
		'card_was_flipped': card_was_flipped,
		'time_up':          time_up,
		
		'time_taken': time_taken,
		
		'start_time':   start_time,
		'use_time':     use_time,
		'time_limit':   time_limit,
		'randomize':    randomized,
		'use_reverse':  use_reverse,
		'random_queue': random_queue,
		'repeat_queue': repeat_queue,
		
		'current_item': current_item,
		'last_item':    last_item
	}, context_instance=RequestContext(request))

File: demo/pages/static/js/unminified/study.js
(function($) {
	var content = $('#wrap'),
		isStudyPage = content.hasClass('study-page');
	
	if (isStudyPage) {
		var cardFront = content.find('.card-front');
		
		// Allow the front of the card to be centered vertically
		if (cardFront.length !== 0) {
			cardFront.parent().css('line-height', cardFront.parent().height()+'px');
		}
		
		
		
		// If progress is being saved, we can provide a better studying experience
		if (Sleepless.storageSet) {
			var form  = $('#study-form'),
				front = $('#front'),
				back  = $('#back'),
				
				
				elements = {
					// Card elements
					card: {
						front: front,
						back:  back.find('.card'),
						fields: {
							front: $('.front-field'),
							back:  $('.back-field'),
							extra: $('.extra-field')
						}
					},
					
					// Form elements
					currentItem: form.find('input[name=current-item]'),
					lastItem:    form.find('input[name=last-item]'),
					
					// Status elements
					status: {
						container:   $('.study-status'),
						currentItem: $('.study-progress .current-item'),
						lastItem:    $('.study-progress .last-item'),
						repeatCount: $('.study-repeats')
					}
				},
				
				
				/**
				* Updates the card fields in the DOM with the given card data.
				* @param hash cardData A hash with the front, back, and extra card data to insert into the DOM.
				*/
				updateCardFields = function(cardData) {
					// @TODO 3 reflows, this could be better
					elements.card.fields.front.text(cardData.front);
					elements.card.fields.back.text(cardData.back);
					elements.card.fields.extra.text(cardData.extra);
				},
				
				
				/**
				* Updates the repeated card counter in the DOM.
				*/
				updateRepeatCounter = function() {
					// If the repeated card counter is not currently in the DOM, we'll have to create it.
					if (!elements.status.repeatCount.length) {
						elements.status.repeatCount = $('<span class="study-status-item study-repeats" />').appendTo(elements.status.container);
					}
					
					// Update the repeated card counter if there are cards in the repeat queue
					if (repeatQueue.length) {
						elements.status.repeatCount.text(repeatQueue.length);
					// Otherwise, remove it
					} else {
						elements.status.repeatCount.remove();
						elements.status.repeatCount = $();
					}
				},
				
				
				/**
				* Returns an array of the number of days each option should delay the card's review by.
				* @param number numberOfOptions The number of delay options that are available to the user.
				* @param number delayFactor The delay factor for this card.
				* @return array An array of the number of days each option should delay by
				*/
				getDelayTimes = function(numberOfOptions, delayFactor) {
					times = [];
					
					for (var i = 0; i < numberOfOptions; i++) {
						times.push(Math.ceil(Math.pow(delayFactor + 2, i + 1 - (i*0.5))));
					}
					
					return times;
				},
				
				
				/**
				* Returns the delay time as a human-readable HTML string.
				* @param number days The number of days that delay would be delayed by.
				* @return string A human readable HTML string.
				*/
				getDelayTimeString = function(days) {
					var numeral = '<span class="button-numeral">NUMERAL</span>',
						time;
					
					// Display the time in terms of days
					if (days < 31) {
						numeral = numeral.replace('NUMERAL', days);
						return interpolate(ngettext('%s day', '%s days', days), [numeral]);
						
					// Display the time in terms of months
					} else if (days < 365) {
						time = Math.round( days / 31 * 10 ) / 10; // Round to 0 or 1 decimal places
						
						numeral = numeral.replace('NUMERAL', time);
						return interpolate(ngettext('%s month', '%s months', time), [numeral]);
						
					// Display the time in terms of years
					} else {
						time = Math.round( days / 365 * 10 ) / 10; // Round to 0 or 1 decimal places
						
						numeral = numeral.replace('NUMERAL', time);
						return interpolate(ngettext('%s year', '%s years', time), [numeral]);
					}
				},
				
				
				startTime   = new Date().getTime() / 1000,
				currentItem = 0,
				lastItem    = 20,
				repeatQueue = [],
				currentCardID,
				currentCard,
				cards,
				
				
				deckID    = content.find('.deck-id').val(),
				cardCount = content.find('.card-count').val(),
				
				
				// Load our progress in this deck
				progress = Sleepless.loadProgress(deckID, {loadCardForecast: true, forceReview: true});
			
			
			
			// If progress hasn't been saved yet, initialize it
			if (progress === undefined) {
				Sleepless.saveProgress(Sleepless.initializeProgress(deckID, cardCount));
				progress = Sleepless.loadProgress(deckID, {loadCardForecast: true, forceReview: true});
			}
			
			// Update the options for this deck
			progress.deck.useTime    = !!form.find('input[name=use-time]').val();
			progress.deck.timeLimit  = parseInt(form.find('input[name=time-limit]').val(), 10) / 60;
			progress.deck.randomize  = !!form.find('input[name=randomize]').val();
			progress.deck.useReverse = !!form.find('input[name=use-reverse]').val();
			progress.deck.cardsPerSession = parseInt(form.find('input[name=cards-per-session]').val(), 10);
			progress.deck.timesStudiedToday++;
			progress.deck.lastStudyDate = new Date(); // Set the last study date to today
			
			Sleepless.saveProgress(progress); // Save changes to the user's progress
			
			
			
			// Add review delay form elements to the DOM
			$('<button type="submit" class="btn bad-btn large-btn" name="repeat" value="1" id="repeat">'+gettext('Repeat')+'</button>').hide().appendTo(form);
			
			$('<div class="good-review-delay-options" />').append(
				$('<p class="review-delay-caption">'+gettext('Review in')+'</p>')
			).append(
				$('<button type="submit" class="btn good-btn large-btn" name="short-review-delay" value="1" />')
			).append(
				$('<button type="submit" class="btn good-btn large-btn" name="medium-review-delay" value="1" />')
			).append(
				$('<button type="submit" class="btn good-btn large-btn" name="long-review-delay" value="1" />')
			).hide().appendTo(form);
			
			elements.showAnswer = form.find('#show-answer');
			// @TODO This is kind of bad
			elements.reviewOptions = {
				container:   form.find('#repeat, .good-review-delay-options'),
				buttons:     $('#repeat, .good-review-delay-options button'),
				goodButtons: $('.good-review-delay-options').find('button')
			};
			
			
			
			// Hide pertinent UI elements until our card data loads in
			elements.status.container.hide();
			elements.card.front.hide();
			elements.showAnswer.hide();
			
			
			// Fetch the card data we'll need for this session
			$.getJSON('/decks/'+deckID+'/cards/', {
				total:   progress.deck.cardsPerSession,
				studied: Sleepless.encodeNumericArray(progress.cards.forecast.studied),
				include: Sleepless.encodeNumericArray(progress.cards.forecast.review),
				randomize:     progress.deck.randomize,
				randomizeSeed: progress.deck.randomizeSeed
			}, function(data) {
				cards = data
				
				
				// Set up our session variables
				lastItem = data.length;
				elements.status.lastItem.text(lastItem);
				elements.lastItem.val(lastItem);
				
				
				// @TODO Write a method for this
				currentCardDataIndex = 0; // Current index in our card data array
				currentCardID = cards[0][0]; // First item is the card's order in the deck
				currentCard = cards[0][1]; // Second item is a hash of card data
				updateCardFields(currentCard);
				
				elements.status.container.show();
				elements.card.front.show(); // Show the front of the current card
				elements.showAnswer.show(); // Show the "show answer" button
				
				
				// Show the other side of the card
				elements.showAnswer.on('click', function(e) {
					var times;
					
					e.preventDefault();
					// @TODO Keyboard shortcuts
					// Swap the form controls
					elements.showAnswer.hide();
					elements.reviewOptions.container.show();
					
					// Show the back of the card
					elements.card.back.show();
					
					
					// Present the user with delay options for when to next review this card
					times = getDelayTimes(elements.reviewOptions.goodButtons.length, progress.cards[currentCardID].delayFactor);
					elements.reviewOptions.goodButtons.each(function(i) {
						$(this).val(times[i]).html(getDelayTimeString(times[i]));
					});
					
					
					// Update our current item if it's not also our last item
					if (currentItem !== lastItem) {
						currentItem++;
						elements.currentItem.val(currentItem);
						elements.status.currentItem.text(currentItem);
					// If we have cards in the repeat queue, we should remove the first item now
					} else if (repeatQueue.length) {
						repeatQueue.shift(); // We have to remove the item here to prevent the session from ending prematurely
						
						// Update the repeated card counter in the DOM
						updateRepeatCounter();
					}
				});
				
				// Show the next card
				elements.reviewOptions.buttons.on('click', function(e) {
					var button = $(e.currentTarget);
					
					e.preventDefault();
					
					// @TODO Keyboard shortcuts
					// Swap the form controls
					elements.showAnswer.show();
					elements.reviewOptions.container.hide();
					
					// Hide the back of the card
					elements.card.back.hide();
					
					
					progress.cards[currentCardID].studied = true;
					
					// If the repeat button was clicked
					if (button.attr('name') === 'repeat') {
						// Reset the delay factor for this card
						progress.cards[currentCardID].delayFactor = 0;
						
						// Add this card's index in our card data array to the repeat queue
						repeatQueue.push(currentCardDataIndex);
						
						// Update the repeated card counter in the DOM
						updateRepeatCounter();
					// If a review delay button was clicked
					} else {
						// Set the card's next review date
						progress.cards[currentCardID].reviewDate.setDate(new Date().getDate() + parseInt(button.val(), 10));
						
						// Increment the card's delayFactor
						progress.cards[currentCardID].delayFactor += 1;
					}
					
					
					// Save our progress
					Sleepless.saveProgress(progress);
					
					
					// If the time limit is up, submit the form
					if (progress.deck.useTime && (new Date().getTime() / 1000 - startTime) / 60 >= progress.deck.timeLimit) {
						form.submit();
						return;
					}
					
					// If this is the last item and we don't have a repeat queue, submit the form
					if (currentItem === lastItem && !repeatQueue.length) {
						form.submit();
						return;
					}
					
					
					
					// If the current item is not the last item, our next card should come from the regular queue
					if (currentItem !== lastItem) {
						currentCardDataIndex = currentItem; // Current index in our card data array
						currentCardID = cards[currentItem][0]; // First item is the card's order in the deck
						currentCard = cards[currentItem][1]; // Second item is a hash of card data
					// Otherwise, the next card should come from the repeat queue
					} else {
						// Get the value of the next card's order from the repeat queue, but don't remove it from the queue
						// This prevents the session from ending prematurely after the user chooses "show answer" on the last repeated card
						currentCardDataIndex = repeatQueue[0]; // Current index in our card data array
						currentCardID = cards[repeatQueue[0]][0]; // First item is the card's order in the deck
						currentCard = cards[repeatQueue[0]][1]; // Second item is a hash of card data
					}
					
					// Inject the new card into the DOM
					updateCardFields(currentCard);
				});
			});
		}
	}
}(jQuery));